### PR TITLE
[ICTL-975] Adjusting the plugin structure to enable different languages dependency

### PR DESCRIPTION
**Pull request by Frosendroska**
# Description of changes made

The full description of changes with the necessary links can be found in [the YouTrack ticket](https://youtrack.jetbrains.com/issue/ICTL-975/Adjusting-the-plugin-structure-to-enable-different-languages-dependency).

## Summary 

**Project Structure Adjustment:**
- Creation of separate modules for Java and Kotlin.
- A common module (langwrappers) with shared interfaces (PsiHelper, PsiHelperProvider).

**Implementation Details:**

- **langwrappers:** Contains `PsiHelper interface` and `PsiHelperProvider class` (EP).
- **java:** Implements Java-specific `PsiHelper` and `PsiHelperProvider` (EP implementation).
- **kotlin:** Implements Kotlin-specific `PsiHelper` and `PsiHelperProvider` (EP implementation).
The main plugin configuration file declares the extension point for language support, allowing the appropriate `PsiHelperProvider` class to be loaded based on the project's language.

# Why is merge request needed

The PR aims to supporting multiple programming languages. The plugin's structure is being modularized to manage dependencies more efficiently, ensuring only necessary language dependencies are included, preventing errors when unsupported languages are used, and facilitating incremental dependencies. As a side-bonus, addition of new languages will be easier.

# Other notes
Closes #ICTL-975

- [x] I have checked that I am merging into correct branch

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -13,6 +13,8 @@ evosuiteVersion = 1.0.5
 pluginSinceBuild = 241
 pluginUntilBuild = 241.*
 
+ideVersionVerifier = IC-2024.1
+
```
Comment (Vladislav0Art) [inline gradle.properties:17]: How is this variable used?

File: java/src/main/kotlin/org/jetbrains/research/testspark/java/JavaPsiHelper.kt [hunk 2]
```diff
@@ -85,10 +84,8 @@ class JavaPsiHelper(private val psiFile: PsiFile) : PsiHelper {
         project: Project,
         classesToTest: MutableList<PsiClassWrapper>,
         caretOffset: Int,
+        maxPolymorphismDepth: Int, // check if cut has any none java super class
```
Comment (Vladislav0Art) [inline java/src/main/kotlin/org/jetbrains/research/testspark/java/JavaPsiHelper.kt:87]: `has any non-java super class`?

File: kotlin/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+tasks {
+    buildSearchableOptions {
+        enabled = false
+    }
+}
+
+kotlin {
+    jvmToolchain(17)
+}
```
Comment (Vladislav0Art) [inline kotlin/build.gradle.kts:34]: This `jvmToolchain(17)` is used multiple times. 

I suggest moving the `17` into a variable on the upper `gradle.build` level and propagating it down to the dependent `gradle.build` scripts.

File: kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiClassWrapper.kt [hunk 4]
```diff
@@ -19,18 +18,18 @@ import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
...
-import org.jetbrains.research.testspark.helpers.psi.PsiClassWrapper
-import org.jetbrains.research.testspark.helpers.psi.PsiMethodWrapper
+import org.jetbrains.research.testspark.langwrappers.PsiClassWrapper
+import org.jetbrains.research.testspark.langwrappers.PsiMethodWrapper
 
 class KotlinPsiClassWrapper(private val psiClass: KtClassOrObject) : PsiClassWrapper {
     override val name: String get() = psiClass.name ?: ""
 
-    override val qualifiedName: String get() = psiClass.fqName!!.asString() ?: ""
+    override val qualifiedName: String get() = psiClass.fqName?.asString() ?: ""
```
Comment (Vladislav0Art) [inline kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiClassWrapper.kt:27]: I remember we discussed not tolerating empty, fully qualified names but checking whether they are null prior he instantiation of `KotlinPsiClassWrapper`? 

Unfortunately, I could not find it. What was the decision, and will it be reflected as an issue?

File: gradle.properties (hunk 1 above)
Comment (Frosendroska) [reply gradle.properties:17]: It does not, thank you! I will delete it)

File: kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiHelper.kt [hunk 5]
```diff
@@ -43,10 +40,12 @@ class KotlinPsiHelper(private val psiFile: PsiFile) : PsiHelper {
         val element = psiFile.findElementAt(caretOffset)
         val cls = element?.parentOfType<KtClassOrObject>(withSelf = true)
 
-        if (cls != null && cls.name != null && cls.fqName != null) {
-            val kotlinClassWrapper = KotlinPsiClassWrapper(cls)
-            log.info("Surrounding class for caret in $caretOffset is ${kotlinClassWrapper.qualifiedName}")
-            return kotlinClassWrapper
+        if (cls != null) {
+            if (cls.name != null) {
```
Comment (Vladislav0Art) [inline kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiHelper.kt:44]: Why not `if(cls != null && cls.name != null)`?

File: kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiHelper.kt [hunk 6]
```diff
@@ -86,8 +87,8 @@ class KotlinPsiHelper(private val psiFile: PsiFile) : PsiHelper {
         project: Project,
         classesToTest: MutableList<PsiClassWrapper>,
         caretOffset: Int,
+        maxPolymorphismDepth: Int, // check if cut has any none java super class
```
Comment (Vladislav0Art) [inline kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiHelper.kt:90]: `non-java super class`

File: kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiMethodWrapper.kt [hunk 7]
```diff
@@ -12,12 +12,12 @@ import org.jetbrains.kotlin.psi.KtPrimaryConstructor
...
 import org.jetbrains.kotlin.psi.psiUtil.containingClassOrObject
-import org.jetbrains.research.testspark.helpers.psi.PsiClassWrapper
-import org.jetbrains.research.testspark.helpers.psi.PsiMethodWrapper
+import org.jetbrains.research.testspark.langwrappers.PsiClassWrapper
+import org.jetbrains.research.testspark.langwrappers.PsiMethodWrapper
 
 class KotlinPsiMethodWrapper(val psiFunction: KtFunction) : PsiMethodWrapper {
 
-    override val name: String get() = psiFunction.name!!
+    override val name: String get() = psiFunction.name ?: ""
```
Comment (Vladislav0Art) [inline kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiMethodWrapper.kt:20]: Here also the question about nullable names

File: kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiMethodWrapper.kt [hunk 8]
```diff
@@ -51,19 +51,14 @@ class KotlinPsiMethodWrapper(val psiFunction: KtFunction) : PsiMethodWrapper {
     val isDefaultMethod: Boolean = psiFunction.run {
         val containingClass = PsiTreeUtil.getParentOfType(this, KtClassOrObject::class.java)
         val containingInterface = containingClass?.isInterfaceClass()
-        // ensure that the function is a non-abstract method defined in an interface
-        name != "<init>" && // function is not a constructor
-            bodyExpression != null && // function has an implementation
-            containingInterface == true // function is defined within an interface
+        name != "<init>" && bodyExpression != null && containingInterface == true
```
Comment (Vladislav0Art) [inline kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiMethodWrapper.kt:54]: Before, there were comments:
```kotlin
       // ensure that the function is a non-abstract method defined in an interface	        name != "<init>" && bodyExpression != null && containingInterface == true
        name != "<init>" && // function is not a constructor	
            bodyExpression != null && // function has an implementation	
            containingInterface == true // function is defined within an interface
```
It was better with comments.

File: kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiMethodWrapper.kt [hunk 9]
```diff
@@ -92,7 +87,6 @@ class KotlinPsiMethodWrapper(val psiFunction: KtFunction) : PsiMethodWrapper {
 
     /**
      * Generates the field descriptor for a type.
-     * https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3
      *
      * @param type the type to generate the descriptor for
```
Comment (Vladislav0Art) [inline kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiMethodWrapper.kt:91]: Here was a link to the docs:
```kotlin
/*
* ...
* https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3
* */
```

File: langwrappers/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    downloadSources.set(true)
+}
+
+tasks.named("verifyPlugin") { enabled = false }
+tasks.named("runIde") { enabled = false }
+tasks.named("runPluginVerifier") { enabled = false }
+
+kotlin {
+    jvmToolchain(17)
+}
```
Comment (Vladislav0Art) [inline langwrappers/build.gradle.kts:29]: Here also a duplication of jvm 17.It should be moved into a variable, e.g., `JVM_TOOLCHAIN_VERSION`.

File: src/main/kotlin/org/jetbrains/research/testspark/actions/TestSparkAction.kt [hunk 11]
```diff
@@ -318,49 +325,79 @@ class TestSparkAction : AnAction() {
...
+                        caretOffset,
+                        fileUrl,
+                        testSamplesCode,
+                        testGenerationController,
+                    )
+                }
+            }
+            visibilityController.isVisible = false
+            dispose()
+        }
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/actions/TestSparkAction.kt:400]: The same duplication got introduced again. Maybe, you have not rebased on the `development`?

File: src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt [hunk 12]
```diff
@@ -210,7 +210,7 @@ object LLMHelper {
      * @return True if the token is set, false otherwise.
      */
     fun isCorrectToken(project: Project, errorMonitor: ErrorMonitor): Boolean {
-        if (!SettingsArguments(project).isTokenSet()) {
+        if (!SettingsArgumentsLlm(project).isTokenSet()) {
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt:213]: It is better to call this class `LlmSettingsArguments`; now it is a bit clumsy.

File: src/main/kotlin/org/jetbrains/research/testspark/services/CoverageVisualisationService.kt [hunk 13]
```diff
@@ -60,7 +60,7 @@ class CoverageVisualisationService(private val project: Project) {
      */
     fun clear() {
         currentHighlightedData ?: return
-        currentHighlightedData!!.editor.markupModel ?: return
+        currentHighlightedData!!.editor.markupModel
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/services/CoverageVisualisationService.kt:63]: Why was the `?: return` removed from here?

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt [hunk 14]
```diff
@@ -40,10 +40,11 @@ class Llm(override val name: String = "LLM") : Tool {
         projectSDKPath: Path? = null,
     ): LLMProcessManager {
         val classesToTest = mutableListOf<PsiClassWrapper>()
+        val maxPolymorphismDepth = SettingsArgumentsLlm(project).maxPolyDepth(0)
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt:43]: Could you make that `0` argument a named argument?

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/PromptManager.kt [hunk 15]
```diff
@@ -264,8 +267,10 @@ class PromptManager(
     ): String {
         for (currentPsiMethod in psiClass.allMethods) {
             if (currentPsiMethod.containsLine(lineNumber)) {
-                val psiHelper = PsiHelperFactory.getPsiHelper(psiClass.containingFile) ?: return ""
-                return psiHelper.generateMethodDescriptor(currentPsiMethod)
+                val file = psiClass.containingFile
+                val psiHelper = PsiHelperProvider.getPsiHelper(file)
+                // psiHelper will not be null here
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/PromptManager.kt:272]: Also, why is not it null here?

File: src/main/resources/META-INF/plugin.xml [hunk 16]
```diff
@@ -1,6 +1,6 @@
 <!-- Plugin Configuration File. Read more: https://plugins.jetbrains.com/docs/intellij/plugin-configuration-file.html -->
 <idea-plugin>
-    <id>org.jetbrains.research.testgenie</id>
+    <id>org.jetbrains.research.testspark</id>
     <name>TestSpark</name>
```
Comment (Vladislav0Art) [inline src/main/resources/META-INF/plugin.xml:4]: Won't it break something on the Plugin Marketplace? What are the consequences of changing the plugin id?

CC: @pderakhshanfar, @arksap2002.

File: src/test/kotlin/org/jetbrains/research/testspark/runner/SettingsArgumentsLlmEvoSuitTest.kt [hunk 17]
```diff
@@ -9,14 +9,14 @@ import com.intellij.testFramework.fixtures.TestFixtureBuilder
...
+import org.jetbrains.research.testspark.tools.evosuite.SettingsArgumentsEvoSuit
 import org.junit.jupiter.api.AfterEach
 import org.junit.jupiter.api.BeforeEach
 import org.junit.jupiter.api.Test
 import org.junit.jupiter.api.TestInstance
 
 @TestInstance(TestInstance.Lifecycle.PER_METHOD)
-class SettingsArgumentsTest {
+class SettingsArgumentsLlmEvoSuitTest {
     private lateinit var settingsState: EvoSuiteSettingsState
```
Comment (Vladislav0Art) [inline src/test/kotlin/org/jetbrains/research/testspark/runner/SettingsArgumentsLlmEvoSuitTest.kt:20]: `SettingsArgumentsLlmEvoSuitTest` -> `SettingsArgumentsLlmEvoSuiteTest`

**Review by Vladislav0Art — CHANGES_REQUESTED** [review]
The most important comment is about the again-introduced duplication in `src/main/kotlin/org/jetbrains/research/testspark/actions/TestSparkAction.kt`. 

I will approve once this comment is resolved and the duplication is removed (likely, you only need to merge/rebase onto the newest `development` branch).

File: kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiClassWrapper.kt (hunk 4 above)
Comment (Frosendroska) [reply kotlin/src/main/kotlin/org/jetbrains/research/testspark/kotlin/KotlinPsiClassWrapper.kt:27]: You are totally correct. It is my bad merge

**Review by Vladislav0Art — APPROVED** [review]

File: src/main/resources/META-INF/plugin.xml (hunk 16 above)
Comment (arksap2002) [reply src/main/resources/META-INF/plugin.xml:4]: Screenshot 2024-06-25 at 2 53 22 AM

I think, we should keep `testgenie`

Comment (pderakhshanfar) [reply src/main/resources/META-INF/plugin.xml:4]: This is our unique ID and should not be changed. There is no way to change it on the marketplace as well. Please keep it as it is

**Review by pderakhshanfar — CHANGES_REQUESTED** [review]
The functionality of the changes are checked and tested and everything looks fine. I also checked the changes. Again, everything looks fine, except the Plugin ID. The ID should not be changed as this is the way that the marketplace recognizes this plugin and

File: src/main/resources/META-INF/plugin.xml (hunk 16 above)
Comment (Frosendroska) [reply src/main/resources/META-INF/plugin.xml:4]: Interestingly, renaming this back + adjusting some obvious things in other `*.xml` files lead to the error that we can not find other `*.xml` files rn.....
 I am working on this.

Comment (Frosendroska) [reply src/main/resources/META-INF/plugin.xml]: fixed

**Review by pderakhshanfar — APPROVED** [review]

**Pull request merged**